rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,327 @@
+There are two main kinds of presets: framework-specific presets, like `babel-preset-react-app`, and company-wide presets, used to share the same plugins accross different projects. Since assumptions describe the input code and not the used plugins, this new option fits better in company-wide presets where the person writing the preset and the person writing the input code have less "degrees of separation", but there are also use cases in framework-specific presets: for example, `babel-preset-react-app` [enables](https://github.com/facebook/create-react-app/blob/c87ab79559e98a5dae2cd0b02477c38ff6113e6a/packages/babel-preset-react-app/create.js#L151-L154) `loose: true` for `@babel/plugin-proposal-class-properties`.
+
+To avoid conflicts between assumptions set in presets, they will only be able to _enable_ them (i.e. set them to `true`). The only way to disable an already enabled assumption is to explicitly do it in a configuration file or in programmatic options. ([discussion](https://github.com/babel/rfcs/pull/5#discussion_r507979771))
+
+## Configuration merging
+
+Multiple `"assumptions"` objects should be merged using `Object.assign`, and not overwritten like other options.
+This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.
+
+They should be merged considering the following precedence, which is the same as what is already used for other options:
+- preset &lt; `babel.config.json` &lt; `.babelrc.json` (far from the compiled file)&lt; `.babelrc.json` (near to the compiled file) &lt; programmatic options
+- top-level &lt; `env` section &lt; `overrides` &lt; `overrides.env`
+
+Assumptions set inside a preset are not boxed inside the preset but are applied to all the plugins, for two reasons:
+1. conceptually, the assumptions describe the input file and not a specific plugin enabled inside the preset;
+2. practically, this makes it possible to have a single "personal" or "company" preset containing assumptions which apply to a user's coding practices, and re-use it in different projects without duplicating the list of assumptions.
+
+In order to expose assumptions defined inside presets to every plugin, we need to first resolve and instantiate all the presets, and then all the plugins with the resolved assumptions. This also means that we cannot provide the defined assumptions list to the presets, because it hasn't been finalized yet when they are instantiated.
+This is implemented by [babel/babel#11689](https://github.com/babel/babel/pull/11689)).
+
+## Plugin API
+
+The first parameter passed to the plugins (often known as `api`) should have a new method: `assumptions(name: string): boolean`, which returns whether or not the assumption has been enabled. If a plugin is asking about an assumption not supported by the used `@babel/core` version, it will return `false`.
```

Comment:
```suggestion
The first parameter passed to the plugins (often known as `api`) should have a new method: `assumptions(name: string): boolean | undefined`, which returns whether or not the assumption has been enabled. If a plugin is asking about an assumption not enabled or not supported by the used `@babel/core` version, it will return `undefined`.
```

(idea from @jridgewell: https://github.com/babel/babel/pull/12407#discussion_r536511944)

By returning `undefined` for assumptions not explicitly set to `false`, plugins can customize their default behavior:
1. They can use `options.loose` as the default value of the assumption they are using.
1. Currently we are hard-coding `newableArrowFunctions` to `true` in `@babel/core` for backward compatibility, but we could move the default value to the plugin itself.

In general, I don't like making assumptions default to `true` in plugins (all assumptions should default to `false`, so that users don't have to also learn about the default values), but I think that if there is an explicit option it's ok.

What do you think?